google/cloud/spanner_dbapi/cursor.py

```diff
@@ -258,9 +258,26 @@ def executemany(self, operation, seq_of_params):
 
         many_result_set = StreamedManyResultSets()
 
-        for params in seq_of_params:
-            self.execute(operation, params)
-            many_result_set.add_iter(self._itr)
+        if classification in (parse_utils.STMT_INSERT, parse_utils.STMT_UPDATING):
+            statements = []
+
+            for params in seq_of_params:
+                sql, params = parse_utils.sql_pyformat_args_to_spanner(
+                    operation, params
+                )
+                statements.append((sql, params, get_param_types(params)))
+
+            transaction = self.connection.transaction_checkout()
+            _, res = transaction.batch_update(statements)
```

Comment:
The usage of `batch_update` is a little bit tricky, and you need to check the status code here. The semantics are as follows:
1. It could be that the `batch_update` function fails with an error. That will be handled like any other error, so there is no special action required for that.
2. It could also be that one of the statements in the list of statements returns an error. If for example your list of statements has 4 statements. If statement number 3 fails, the return value of `batch_update` will be the status code of the error + the update count of the first 2 statements that did succeed (so the length of the second return value should be 2). The 4th statement will not be executed.
3. The error code that is returned as the first return value of this function can also be `Aborted`. Using the same example as above: This means that the first 2 statements executed successfully, but that during the execution of the third statement the transaction was aborted by Cloud Spanner. This should be handled as if the entire statement aborted the transaction (so it should cause a retry of the transaction).

The DbApi specification states for the `executemany` function that the `Return values are not defined.`, so we are free to choose what to return in case an error code is returned by Spanner. The best might be to return the same as Cloud Spanner, which means both the error code and the update counts, except when the error code is `Aborted`. If the error code is `Aborted`, then it's (in my opinion) better to throw an `AbortedError`, as there is nothing more that a user can do with the transaction anyways, and the entire transaction has already been rolled back.

Reply:
The thing is that we don't ever return anything from `execute()` and `executemany()` methods. If one wants to get rows, they should use `fetch*()` methods after `executemany()` method finished. In our case it also will be possible as we're adding DML batch result into cursor results object.